### PR TITLE
Enable using custom PVC without a template for the statefulset (for the buffer volume and also as extravolumes)

### PR DESCRIPTION
**Pull request by bbyreddy**
Ensured that the custom PVC source details specified in the fluentd.extraVolumes.volume.pvc.source.claimName of the fluentd definition are honoured instead of always attempting to create a new PVC.

Introduced empty checks to the PersistentVolumeClaim.PersistentVolumeClaimSpec and PersistentVolumeClaim.PersistentVolumeSource structures and ApplyPVCForStatefulSet() and ApplyVolumeForPodSpec() functions are invoked only when the respective structures are not empty.

### CODE REVIEW

File: pkg/resources/fluentd/statefulset.go [hunk 1]
```diff
@@ -72,6 +77,16 @@ func (r *Reconciler) statefulset() (runtime.Object, reconciler.DesiredState, err
 	return desired, reconciler.StatePresent, nil
 }
 
+func isPersistentVolumeClaimSpecEmpty(pvcSpec corev1.PersistentVolumeClaimSpec) bool {
+	searilizedPvcSpec, _ := json.Marshal(pvcSpec) // Json serialization should not fail, hence the error is ignored
+	return string(searilizedPvcSpec) == `{"resources":{}}`
```
Comment (aslafy-z) [inline pkg/resources/fluentd/statefulset.go:82]:
```suggestion
	serilizedPvcSpec, _ := json.Marshal(pvcSpec) // Json serialization should not fail, hence the error is ignored
	return string(serilizedPvcSpec) == `{"resources":{}}`
```

File: pkg/resources/fluentd/statefulset.go [hunk 2]
```diff
@@ -72,6 +77,16 @@ func (r *Reconciler) statefulset() (runtime.Object, reconciler.DesiredState, err
...
+}
+
+func isPersistentVolumeSourceEmpty(pvcSource corev1.PersistentVolumeClaimVolumeSource) bool {
+	searilizedPvcSource, _ := json.Marshal(pvcSource) // Json serialization should not fail, hence the error is ignored
+	return string(searilizedPvcSource) == `{}`
```
Comment (aslafy-z) [inline pkg/resources/fluentd/statefulset.go:87]:
```suggestion
	serilizedPvcSource, _ := json.Marshal(pvcSource) // Json serialization should not fail, hence the error is ignored
	return string(serilizedPvcSource) == `{}`
```

**Review by aslafy-z — CHANGES_REQUESTED** [review]
I spot a couple of typos, can you please fix them?

**Comment by pepov** [comment]
I promised to look into how we could make the comparison more efficient, sorry I haven't been able to get back to that

**Comment by bbyreddy** [comment]
> I promised to look into how we could make the comparison more efficient, sorry I haven't been able to get back to that

Hi @pepov and @aslafy-z ,
There is another PR https://github.com/kube-logging/logging-operator/pull/1622 created by @OverOrion , which resolves this issue. I'm not sure which direction you want to take to resolve this issue.
Thanks!

File: pkg/resources/fluentd/statefulset.go (hunk 2 above)
Comment (bbyreddy) [reply pkg/resources/fluentd/statefulset.go]: Corrected the spelling.

**Comment by pepov** [comment]
@bbyreddy we want to keep this instead of introducing yet another huge field to the fluentd spec

**Comment by pepov** [comment]
I'm planning to add some technical docs as well

**Comment by bbyreddy** [comment]
> @bbyreddy can you check if the solution that works for you?

@pepov , yes, this works perfectly fine and is much simpler as well. 
Please let me know if I need to do anything to get this PR merged into the master.
Thanks!

**Review by pepov — APPROVED** [review]

**Comment by bbyreddy** [comment]
> I spotted a couple of typos, can you please fix them? Also, can you please update the PR name so it reflects the changes and add `fixes #1616` in the PR description? You will also have to sign your commits, see https://github.com/kube-logging/logging-operator/pull/1617/checks?check_run_id=20378816523.

@aslafy-z , you may have to mark this as resolved and approve the PR as the code has changed now and doesn't have the typos anymore. Thanks!

**Review by OverOrion — APPROVED** [review]

**Comment by pepov** [comment]
@aslafy-z typos has been fixed and signoff done, I'm merging this

**Pull request merged**